internal/gw/client.go

```diff
@@ -145,40 +146,97 @@ func (c *client) uploadBlob(ctx context.Context, item walk.SyncItem) error {
 	return nil
 }
 
+type uploadResult struct {
+	Error error
+	Item  walk.SyncItem
+}
+
+func (c *client) uploadWorker(
+	ctx context.Context,
+	items <-chan walk.SyncItem,
+	results chan<- uploadResult,
+	wg *sync.WaitGroup,
+) {
+	for item := range items {
+		if err := c.uploadBlob(ctx, item); err != nil {
+			results <- uploadResult{err, item}
+			break
+		}
+		results <- uploadResult{nil, item}
+	}
+	wg.Done()
+}
+
 func (c *client) EnsureUploaded(
 	ctx context.Context,
 	items []walk.SyncItem,
 	onUploaded func(walk.SyncItem) error,
 	onPresent func(walk.SyncItem) error,
+	onDuplicate func(walk.SyncItem) error,
 ) error {
-	// TODO: concurrency
+	// Maintain a map of items processed thus far
+	processedItems := make(map[string]walk.SyncItem)
+	// The final collection of items to upload
+	var toUpload []walk.SyncItem
+
+	numThreads := c.cfg.UploadThreads()
+	var wg sync.WaitGroup
+	results := make(chan uploadResult, len(items))
+	jobs := make(chan walk.SyncItem, len(items))
+
+	for i := 0; i < numThreads; i++ {
+		wg.Add(1)
+		go c.uploadWorker(ctx, jobs, results, &wg)
+	}
+
 	for _, item := range items {
 		if item.Key == "" && item.LinkTo != "" {
 			log.FromContext(ctx).F("uri", item.SrcPath).Debug("Skipping unfollowed symlink")
 			continue
 		}
 
-		// Check if it's present
+		// Determine if the blob is already present in the bucket
 		have, err := c.haveBlob(ctx, item)
 		if err != nil {
 			return fmt.Errorf("checking for presence of %s: %w", item.Key, err)
 		}
-
+		// If so, do not include it in the final set of items to upload
 		if have {
 			if err = onPresent(item); err != nil {
 				return err
 			}
 			continue
 		}
 
-		if err = c.uploadBlob(ctx, item); err != nil {
-			return err
+		// Determine if the item already exists in the final set of items to upload
+		// If so, prevent a costly re-upload of the item
+		if _, ok := processedItems[item.Key]; ok {
+			log.FromContext(ctx).F("uri", item.SrcPath).Debug("Skipping duplicate item")
+			if err := onDuplicate(item); err != nil {
+				return err
+			}
+			continue
 		}
-		if err = onUploaded(item); err != nil {
+		processedItems[item.Key] = item
+		jobs <- item
+		toUpload = append(toUpload, item)
```

Comment:
We don't really need toUpload any more, do we?  It seems like the only use of it is to later do a traditional indexed for loop using its length, but couldn't we just loop over the results channel instead?  Though you would have to move the close(results) prior to the loop, which it probably should be anyway.  (reading from a channel after close is fine, and the wg.Wait() should guarantee that no more writes to it are expected)